pyteal/ast/router.py

```diff
@@ -715,12 +1000,31 @@ def build_program(self) -> tuple[Expr, Expr, sdk_abi.Contract]:
             * clear_state_program: an AST for clear-state program
             * contract: a Python SDK Contract object to allow clients to make off-chain calls
         """
+        if not self.bare_call_actions.is_empty():
+            bare_call_approval = self.bare_call_actions.approval_construction()
+            if bare_call_approval:
+                self.approval_ast.bare_calls = [
+                    CondNode(
+                        Txn.application_args.length() == Int(0),
+                        cast(Expr, bare_call_approval),
+                    )
+                ]
+
+        optimize = optimize or OptimizeOptions()
+        use_frame_pt = optimize.use_frame_pointers(version)
         return (
-            self.approval_ast.program_construction(),
+            self.approval_ast.program_construction(use_frame_pt=use_frame_pt),
             self.clear_state,
             self.contract_construct(),
         )
 
+    @contextmanager
+    def _cleaning_context(self):
+        starting_slot_id = ScratchSlot.nextSlotId
+        yield
+        self._clean()
+        ScratchSlot.reset_slot_numbering(starting_slot_id)
```

Comment:
```suggestion
        try:
            yield
        finally:
            self._clean()
            ScratchSlot.reset_slot_numbering(starting_slot_id)
```

After playing around with my own context manager, I believe this pattern may be safer, in case of an exception during the compilation.

Reply:
thanks for the suggestion!